### PR TITLE
Adds gitlab documentation

### PR DESCRIPTION
**Pull request by catenacyber**
For https://github.com/google/oss-fuzz/pull/7073
cc @jonathanmetzman

### CODE REVIEW

**Review by securitykernel — CHANGES_REQUESTED** [review]

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:93]: indeed

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:73]: Leaving this to Johnathan

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:67]: Interesting, I need these on some hosted gitlab.
Changing this in next commit

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:50]: Indeed.
I wanted to have a minimal config for this doc first.
So adding this tip afterwards

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:52]: You are right, thanks for spotting this. Removing this from the doc

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [inline docs/running-clusterfuzzlite/gitlab.md:203]: Thanks for going through the trouble of setting this up.
I have some high level comments.
1. This section seems to offer two options: 1. ??? and 2. using  a personal access token.
Can you tell me the tradeoff between the two and we will decide which one to recommend and then we won't even mention the other option.
2. In the github docs we have screenshots for setting up a personal acccess token (https://google.github.io/clusterfuzzlite/running-clusterfuzzlite/github-actions/#storage-repo). If we go that route, can you add screenshots here?
3. Can you add screenshots and explanation for downloading artifacts please?
4. This section makes artifacts seem optional, why should they be optional?
5. The reason we use the git filestore on github in addition to actions is two-fold: 1. it makes browsing coverage reports on the web possible. 2. It makes it possible for two batch fuzzing jobs to save to the corpus at once (side note: @oliverchang: have we ever tested this? Does batch fuzzing force push so that committing works even if its behind HEAD?). Should we do the same on gitlab?

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [inline docs/running-clusterfuzzlite/gitlab.md:122]: Maybe have some screenshots walk through an example of setting this up.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [reply docs/running-clusterfuzzlite/gitlab.md:68]: The images are exactly the same (for now) except for their entrypoints.
I'm not sure this is something I want to support and I don't like the fact that this is using build_fuzzers for running fuzzers.
I can publish a cifuzz-combined image that uses the combined entrypoint to support both.
but let me explain why we have different images.
On github, having two images makes it easy to seperate build logs from run logs.
See how in [these logs](https://github.com/systemd/systemd/runs/4704498147?check_suite_focus=true) you can click to view the build step and run step seperately.
Now, the drawback of this approach is that I think it has made the interface more complex, instead of having to invoke Cifuzz once it must be invoked twice. See how in this example two steps must be specified: https://google.github.io/clusterfuzzlite/running-clusterfuzzlite/github-actions/#pr-fuzzing

In hindsight, I think it might be a small mistake to have seperated the two actions. So the question is, should we have separate actions on gitlab as well (which will make things consistent at least?)

@oliverchang WDYT of all this?

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [inline docs/running-clusterfuzzlite/gitlab.md:6]: Make this nav order 2 please and make google cloud build 3 and prow 4.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [inline docs/running-clusterfuzzlite/gitlab.md:1]: Can you add a todo to this page to add a link to this doc [here](https://github.com/google/clusterfuzzlite/blob/85bb71ba22c90698f38e01171b6b7efd46ac47ec/docs/clusterfuzzlite.md?plain=1#L41) [here],(https://github.com/google/clusterfuzzlite/blob/1583ba6edd86e16bd57bf85f3dea93e813335972/docs/running_clusterfuzzlite.md?plain=1#L181) and a mention [here](https://github.com/google/clusterfuzzlite/blob/85bb71ba22c90698f38e01171b6b7efd46ac47ec/README.md?plain=1#L34), once this page is finalized?

**Review by jonathanmetzman — COMMENTED** [review]
Thanks!

File: docs/running-clusterfuzzlite/gitlab.md
Comment (oliverchang) [reply docs/running-clusterfuzzlite/gitlab.md:203]: re parallel pushing: we auto rebase and retry when push fails to make this work.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:50]: > nit but are the quotes needed here?

Indeed not

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:27]: ok

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:29]: @securitykernel managed to get Docker-in-Docker to work.
I did not yet, having a problem with setting `DOCKER_HOST` to the right value on my setup...

So, removing this, and hoping to add how to setup with Docker-in-Docker

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:59]: ok

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:64]: > Actually, this comment seems like it's for implementors (us) and not developers maybe drop it?

indeed it is for us ;-)

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:67]: ok

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:70]: `always` means `always but when there is something to upload`
It seems simpler for me : this way we use consistently `always` for all the builds.
For batch fuzzing, we need always : because we need crashes on failures, and corpus on success

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:85]: ok

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:80]: done

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:82]: Good question :-)
Suricata's use case is to use a private branch with all bug fixes which will get into next release
Let me know if you need more details

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:83]: I guess you can have in the future a `FILESTORE` ftp or rsync or whatever...
I agree that the explanation is not complete. How would you caption it ?

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:89]: Because both functions update the corpus.

In Gitlab, there is only one gitlab-ci.yml (which has different jobs), and the schedules are defined independently.
The way to know which schedule triggered is to use some variables.
The job is the same for these 2 functions.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:133]: changing to the proposed wording

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:148]: Maybe you want another doc section about `CFL_BRANCH` ?

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:141]: This is only needed with the gitlab artifacts filestore.

When ClusterFuzzLite Gitlab's artifacts filestore uses Gitlab API, it looks for a job with this specific name `clusterfuzzlite-corpus`
Maybe we could make it configurable instead of hardcoded.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:199]: ok

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:221]: ok

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:216]: ok

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:122]: I can do that once the rest is good for you

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:203]: > Can you tell me the tradeoff between the two and we will decide which one to recommend and then we won't even mention the other option.

The downside of the artifacts is that you need to set up an API token
The downside of the cache is that you should ensure that runners share the cache

I think I gather that you want in the end a `gitlab` filestore that will
- push crashes as jobs artifacts
- push and pull coverage reports in another git repo
- push and pull builds from a cache
- push and pull corpus from a cache

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:6]: ok

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:1]: > Can you add a todo to this page ?

What do you mean ?
I added the links

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:84]: done and it works

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [reply docs/running-clusterfuzzlite/gitlab.md]: Why do we want docker-in-docker?
1. I think it's slower.
2. It's different from how things are done on GH, so it's not the main supported way we have of doing things.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [reply docs/running-clusterfuzzlite/gitlab.md]: Why use cache and artifacts? Why not just one?
Why use git repo for reports? Does gitlab have something like github pages?
Sorry if these are dumb questions/answered elsewhere.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [reply docs/running-clusterfuzzlite/gitlab.md:120]: Sure!

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md]: I am not sure we want it.
It is one of the 3 supported modes to use docker in Gitlab CI cf https://docs.gitlab.com/ee/ci/docker/using_docker_build.html

File: docs/running-clusterfuzzlite/gitlab.md
Comment (securitykernel) [reply docs/running-clusterfuzzlite/gitlab.md:139]: Hardcoding job names seems not very intuitive. Developers use any sorts of job names. If necessary, at least it should be configurable.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (securitykernel) [reply docs/running-clusterfuzzlite/gitlab.md:68]: I see your point now. And I agree that it looks nicer in the logs. In GitLab the log view is different. You would have to specify two jobs per sanitizer, effectively it would look like this (shows 8 jobs):

image

A click on each job leads to the job log.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (securitykernel) [reply docs/running-clusterfuzzlite/gitlab.md:89]: I can see your point from a technical view on the jobs being identical. But from a higher-level point of view, these are different things. So I still would like to see two jobs, one for batch fuzzing (these could be three already with all three sanitizers), and one for pruning. Using the same name for the four jobs is misleading IMHO.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:139]: jobs names are no longer required to be hard coded

**Comment by catenacyber** [comment]
@jonathanmetzman you can take a look at GitLab Pages output here :
https://catenacyber.gitlab.io/suricata-cfl/coverage/latest/report/linux/report.html

Is it good to have everything public or should we restrict it to what is in `coverage/latest/report/` ?

If everything is good for you, I can make the screenshots

File: docs/clusterfuzzlite.md
Comment (jonathanmetzman) [inline docs/clusterfuzzlite.md:41]: This link doesn't go anywhere. Add target to line 79.

File: docs/running_clusterfuzzlite.md
Comment (jonathanmetzman) [inline docs/running_clusterfuzzlite.md:181]: This link doesn't go anywhere. Add it to line 192.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [reply docs/running-clusterfuzzlite/gitlab.md:68]: Not sure. As I mentioned, I think 2 actions/images makes the interface for users more complicated, but I guess that interface is only used once while the seperation provides benefits that are experienced many times.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [reply docs/running-clusterfuzzlite/gitlab.md:82]: ACK

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [inline docs/running-clusterfuzzlite/gitlab.md:83]: nit: Remove space before colon.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [inline docs/running-clusterfuzzlite/gitlab.md:201]: get rid of space before colon.

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [inline docs/running-clusterfuzzlite/gitlab.md:231]: get rid of space before colon.

**Review by jonathanmetzman — COMMENTED** [review]
I think this basically lgtm.
Only thing left is:
1. decide if we want to use one image or two
2. fix nits
3. Let https://github.com/google/oss-fuzz/pull/7073 land before merging this.

**Comment by jonathanmetzman** [comment]
> @jonathanmetzman you can take a look at GitLab Pages output here : https://catenacyber.gitlab.io/suricata-cfl/coverage/latest/report/linux/report.html
> 
> Is it good to have everything public or should we restrict it to what is in `coverage/latest/report/` ?
> 
> If everything is good for you, I can make the screenshots

No need to make it private IMO.
Yes, please make screenshots.

File: docs/clusterfuzzlite.md
Comment (catenacyber) [reply docs/clusterfuzzlite.md:41]: indeed

File: docs/running_clusterfuzzlite.md
Comment (catenacyber) [reply docs/running_clusterfuzzlite.md:181]: indeed

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:201]: ok

File: docs/running-clusterfuzzlite/gitlab.md
Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:231]: ok

**Comment by catenacyber** [comment]
The screenshots are available here :
http://catenacyber.fr/gitlab-schedule-mode.png
http://catenacyber.fr/gitlab-project-token.png
http://catenacyber.fr/gitlab-variable-token.png

Otherwise nits are fixed, so should be good.
Waiting for your decision about splitting or not building and running...

File: docs/running-clusterfuzzlite/gitlab.md
Comment (oliverchang) [reply docs/running-clusterfuzzlite/gitlab.md:68]: Sorry for the late reply. I think we should try to be as consistent as we can and keep the two separate entrypoints for gitlab as well. Does this add significantly more complexity to the gitlab integration?

**Comment by jonathanmetzman** [comment]
> 

I've copied these to our bucket can you replace them with:
https://storage.googleapis.com/clusterfuzzlite-public/images/gitlab-schedule-mode.png
https://storage.googleapis.com/clusterfuzzlite-public/images/gitlab-project-token.png
https://storage.googleapis.com/clusterfuzzlite-public/images/gitlab-variable-token.png

**Comment by catenacyber** [comment]
> I've copied these to our bucket can you replace them with:
> https://storage.googleapis.com/clusterfuzzlite-public/images/gitlab-schedule-mode.png
> https://storage.googleapis.com/clusterfuzzlite-public/images/gitlab-project-token.png
> https://storage.googleapis.com/clusterfuzzlite-public/images/gitlab-variable-token.png

The links already pointed to these urls in gitlab.md ;-)

File: docs/running-clusterfuzzlite/gitlab.md
Comment (jonathanmetzman) [reply docs/running-clusterfuzzlite/gitlab.md:68]: +1

@catenacyber can you please do this?

Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:68]: ok will do

Comment (catenacyber) [reply docs/running-clusterfuzzlite/gitlab.md:68]: Looking into this, I think this adds significant complexity, compared to already integrated CI systems.

We need to share the fuzzers built by `clusterfuzzlite-build-fuzzers` with `clusterfuzzlite-run-fuzzers`
As these are 2 different docker images, we need 2 different GitLab jobs. So they do not share any workspace...
To share data between jobs (no matter if they belong to the same pipeline or not), we either need cache or artifacts.

I am trying the solution where `clusterfuzzlite-build-fuzzers` ends with `cp -r $CI_BUILDS_DIR/$CI_JOB_ID/build-out/ $CI_PROJECT_DIR/artifacts/build-out/$SANITIZER` and `clusterfuzzlite-run-fuzzers` begins with `mv $CI_PROJECT_DIR/artifacts/build-out/$SANITIZER $CI_BUILDS_DIR/$CI_JOB_ID/build-out`
Anyways, the `build-out`contents needs to be put in a specific directory under the `$CI_PROJECT_DIR ` aka `project_src_path`. And we need to differentiate between the different sanitizers builds 

This can work also with cache instead of artifacts.
Note that I need to use `$CI_BUILDS_DIR/$CI_JOB_ID` as gitlab does not know `$WORKSPACE`
Another solution would be to bring back the workspace in the project directory... But we would still need to differentiate between the sanitizers...

**Comment by CarpeDiem-CarpeNoctem** [comment]
Hi there,

sorry for being late but while I was working with this integration I had a problem concerning the corpus feature.

maybe It should be mentioned in the docs that the folder name of the external corpus git repository must be "corpus/{fuzztargetname}". It has to follow this structure or the gitlab integration will not find the necessary corpus files.

Best regards

**Comment by jonathanmetzman** [comment]
> Hi there,
> 
> sorry for being late but while I was working with this integration I had a problem concerning the corpus feature.
> 
> maybe It should be mentioned in the docs that the folder name of the external corpus git repository must be "corpus/{fuzztargetname}". It has to follow this structure or the gitlab integration will not find the necessary corpus files.
> 
> Best regards

Is this for manually adding elements to the corpus?

**Comment by catenacyber** [comment]
> > Hi there,
> > sorry for being late but while I was working with this integration I had a problem concerning the corpus feature.
> > maybe It should be mentioned in the docs that the folder name of the external corpus git repository must be "corpus/{fuzztargetname}". It has to follow this structure or the gitlab integration will not find the necessary corpus files.
> > Best regards
> 
> Is this for manually adding elements to the corpus?

You can also use a seed corpus ad on oss-fuzz with a `{fuzztargetname}_seed_corpus.zip` file ;-)

**Pull request merged**
